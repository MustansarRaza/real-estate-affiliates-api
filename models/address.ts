import { Sequelize, Model, DataTypes } from "sequelize";

class Address extends Model {
  public id!: number;
  public locationID!: number | null;
  public countryLocationID!: number | null;
  public regionLocationID!: number | null;
  public cityLocationID!: number | null;
  public localityLocationID!: number | null;
  public phaseLocationID!: number | null;
  public unitNumber!: string | null;

  static setup(sequelize: Sequelize): void {
    this.init(
      {
        id: {
          type: DataTypes.INTEGER,
          allowNull: false,
          primaryKey: true,
          autoIncrement: true,
          field: "address_id",
        },
        locationID: {
          type: DataTypes.INTEGER,
          allowNull: true,
          field: "location_id",
        },
        countryLocationID: {
          type: DataTypes.INTEGER,
          allowNull: true,
          field: "state_id",
        },
        regionLocationID: {
          type: DataTypes.INTEGER,
          allowNull: true,
          field: "region_id",
        },
        cityLocationID: {
          type: DataTypes.INTEGER,
          allowNull: true,
          field: "city_id",
        },
        localityLocationID: {
          type: DataTypes.INTEGER,
          allowNull: true,
          field: "locality_id",
        },
        phaseLocationID: {
          type: DataTypes.INTEGER,
          allowNull: true,
          field: "phase_id",
        },
        unitNumber: {
          type: DataTypes.STRING,
          allowNull: true,
          field: "unit_number",
        },
      },
      {
        sequelize,
        createdAt: false,
        updatedAt: false,
        tableName: "zn_pf_addresses",
      }
    );
  }
}

export default Address;
