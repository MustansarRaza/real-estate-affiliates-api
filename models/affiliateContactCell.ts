import { Sequelize, Model, DataTypes } from "sequelize";

class AffiliateContactCell extends Model {
  public id!: number;
  public contactID!: number;
  public mobilePhoneNumber!: string;
  public createdAt!: Date;
  public updatedAt!: Date;

  static setup(sequelize: Sequelize): void {
    this.init(
      {
        id: {
          type: DataTypes.INTEGER,
          allowNull: false,
          primaryKey: true,
          autoIncrement: true,
          field: "id",
        },
        contactID: {
          type: DataTypes.INTEGER,
          allowNull: false,
          field: "contact_id",
        },
        mobilePhoneNumber: {
          type: DataTypes.STRING,
          allowNull: false,
          field: "cell",
        },
        createdAt: {
          type: DataTypes.DATE,
          allowNull: false,
          field: "created_at",
        },
        updatedAt: {
          type: DataTypes.DATE,
          allowNull: false,
          field: "updated_at",
        },
      },
      {
        sequelize,
        tableName: "affiliate_contact_cells",
      }
    );
  }
}

export default AffiliateContactCell;
