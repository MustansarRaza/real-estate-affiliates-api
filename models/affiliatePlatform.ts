export enum AffiliatePlatform {
  STAFF_APP = "staff_app",
  WEB_APP = "propforce.com",
  MOBILE_APP = "propforce_mobile_app",
}

export enum AffiliatePlatformID {
  STAFF_APP = 1,
  WEB_APP = 2,
  MOBILE_APP = 3,
}

/**
 * Converts from an enum string value representing an affiliate platform into an ID.
 */
export const affiliatePlatformToID = (
  affiliatePlatform: AffiliatePlatform | null
): AffiliatePlatformID | null => {
  switch (affiliatePlatform) {
    case AffiliatePlatform.STAFF_APP:
      return AffiliatePlatformID.STAFF_APP;

    case AffiliatePlatform.WEB_APP:
      return AffiliatePlatformID.WEB_APP;

    case AffiliatePlatform.MOBILE_APP:
      return AffiliatePlatformID.MOBILE_APP;

    default:
      return null;
  }
};
