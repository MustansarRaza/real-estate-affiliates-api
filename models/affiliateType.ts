export enum AffiliateType {
  AGENCY = "agency",
  INDIVIDUAL = "individual",
  BROKER = "broker",
  AGENT = "agent",
  TITANIUM_AGENCY = "titanium_agency",
}

export enum AffiliateTypeID {
  AGENCY = 1,
  INDIVIDUAL = 2,
  BROKER = 3,
  AGENT = 4,
  TITANIUM_AGENCY = 5,
}

/**
 * Converts from an ID representing an affiliate type to an enum string value.
 *
 * This is done because historically, the available affiliates types are stored
 * in the database.
 *
 * The available list of purposes is finite and high unlikely to change
 * over time. Therefor, we gain in performance what we lose in flexibility.
 */
export const affiliateTypeFromID = (
  affiliateTypeID: AffiliateTypeID | null
): AffiliateType | null => {
  switch (affiliateTypeID) {
    case AffiliateTypeID.AGENCY:
      return AffiliateType.AGENCY;

    case AffiliateTypeID.INDIVIDUAL:
      return AffiliateType.INDIVIDUAL;

    case AffiliateTypeID.BROKER:
      return AffiliateType.BROKER;

    case AffiliateTypeID.AGENT:
      return AffiliateType.AGENT;

    case AffiliateTypeID.TITANIUM_AGENCY:
      return AffiliateType.TITANIUM_AGENCY;

    default:
      return null;
  }
};

/**
 * Converts from an enum string value representing an affiliate type into an ID.
 */
export const affiliateTypeToID = (affiliateType: AffiliateType | null): AffiliateTypeID | null => {
  switch (affiliateType) {
    case AffiliateType.AGENCY:
      return AffiliateTypeID.AGENCY;

    case AffiliateType.INDIVIDUAL:
      return AffiliateTypeID.INDIVIDUAL;

    case AffiliateType.BROKER:
      return AffiliateTypeID.BROKER;

    case AffiliateType.AGENT:
      return AffiliateTypeID.AGENT;

    case AffiliateType.TITANIUM_AGENCY:
      return AffiliateTypeID.TITANIUM_AGENCY;

    default:
      return null;
  }
};
