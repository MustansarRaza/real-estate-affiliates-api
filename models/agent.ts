import { DataTypes, Model, Sequelize } from "sequelize";

class Agent extends Model {
  public id!: number;
  public isTest!: boolean;

  static setup(sequelize: Sequelize): void {
    this.init(
      {
        id: {
          type: DataTypes.INTEGER({ length: 11, unsigned: true }),
          primaryKey: true,
          allowNull: false,
          autoIncrement: true,
          field: "agentid",
        },
        isTest: {
          type: DataTypes.BOOLEAN,
          allowNull: false,
          defaultValue: 0,
          field: "is_test",
        },
      },
      {
        sequelize,
        createdAt: false,
        updatedAt: false,
        tableName: "zn_agents",
      }
    );
  }
}

export default Agent;
