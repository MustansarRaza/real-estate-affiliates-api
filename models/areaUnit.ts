export enum AreaUnit {
  SQM = "sqm",
  KANAL = "kanal",
  SQYD = "sqyd",
  SQFT = "sqft",
  MARLA = "marla",
}

export enum AreaUnitID {
  SQM = 1,
  KANAL = 2,
  SQYD = 3,
  SQFT = 4,
  MARLA = 5,
}

/**
 * Converts from an ID representing an area unit to an enum string value.
 *
 * This is done because historically, the available area units are stored
 * in the database. Retrieving the area unit for a listing, project or lead
 * would require yet another join.
 *
 * The available list of area units is finite and high unlikely to change
 * over time. Therefor, we gain in performance what we lose in flexibility.
 */
export const areaUnitFromID = (areaUnitID: AreaUnitID | null): AreaUnit | null => {
  switch (areaUnitID) {
    case AreaUnitID.SQM:
      return AreaUnit.SQM;

    case AreaUnitID.KANAL:
      return AreaUnit.KANAL;

    case AreaUnitID.SQYD:
      return AreaUnit.SQYD;

    case AreaUnitID.SQFT:
      return AreaUnit.SQFT;

    case AreaUnitID.MARLA:
      return AreaUnit.MARLA;

    default:
      return null;
  }
};
