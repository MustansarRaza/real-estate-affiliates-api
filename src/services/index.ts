import AppRelease from "./AppRelease";
import CategoriesService from "./categories";
import ClientsService, { CreateClientDTO as CreateClientDTOType } from "./clients";
import CommissionsService from "./commissions";
import LeadsService, { CreateLeadDTO as CreateLeadDTOType } from "./leads";
import ListingsService, { ListingDTO as ListingDTOType } from "./listings";
import LocationsService from "./locations";
import MeService, { CreateMeDTO as CreateMeDTOType, MeDTO as MeDTOType } from "./me";
import ProfileService from "./profile";
import ProjectsService from "./projects";
import ListingReservationService, { ListingReservationOptions } from "./reservations";
import TasksService from "./tasks";
import TaskTypesService from "./taskTypes";

export interface ServiceMesh {
  profile: ProfileService;
  me: MeService;
  clients: ClientsService;
  tasks: TasksService;
  taskTypes: TaskTypesService;
  leads: LeadsService;
  projects: ProjectsService;
  listings: ListingsService;
  locations: LocationsService;
  categories: CategoriesService;
  reservations: ListingReservationService;
  commissions: CommissionsService;
  appRelease: AppRelease;
}

const useServices = (): ServiceMesh => {
  const locations = new LocationsService();
  const profile = new ProfileService({ locations });
  const me = new MeService({ profile, locations });
  const clients = new ClientsService({ me, locations });
  const taskTypes = new TaskTypesService();
  const leads = new LeadsService({ me, taskTypes });
  const tasks = new TasksService({ leads, taskTypes });
  const listings = new ListingsService({ me });
  const commissions = new CommissionsService();
  const projects = new ProjectsService({ me, listings, commissions, locations });
  const categories = new CategoriesService();
  const reservations = new ListingReservationService({ me, listings });
  const appRelease = new AppRelease();

  return {
    profile,
    me,
    clients,
    leads,
    tasks,
    taskTypes,
    projects,
    listings,
    locations,
    categories,
    reservations,
    commissions,
    appRelease,
  };
};

export {
  useServices,
  ClientsService,
  LeadsService,
  MeService,
  ProfileService,
  ProjectsService,
  ListingsService,
  TasksService,
  TaskTypesService,
  LocationsService,
  CategoriesService,
  ListingReservationService,
  CommissionsService,
  AppRelease,
};

export type CreateClientDTO = CreateClientDTOType;
export type CreateLeadDTO = CreateLeadDTOType;
export type CreateMeDTO = CreateMeDTOType;
export type CreateReservationDTO = ListingReservationOptions;
export type ListingDTO = ListingDTOType;
export type MeDTO = MeDTOType;
