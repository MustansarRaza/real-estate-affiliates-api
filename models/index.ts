import { config } from "affiliates-api/config";
import { getDbTimezone } from "affiliates-api/features";
import cls from "cls-hooked";
import { Sequelize } from "sequelize";
import Address from "./address";
import Affiliate from "./affiliate";
import AffiliateAssignee, { AffiliateAssigneeType } from "./affiliateAssignee";
import AffiliateCommission from "./affiliateCommission";
import AffiliateCommissionPropertyType from "./affiliateCommissionPropertyType";
import AffiliateContact from "./affiliateContact";
import AffiliateContactCell from "./affiliateContactCell";
import { AffiliatePlatform, AffiliatePlatformID, affiliatePlatformToID } from "./affiliatePlatform";
import AffiliateStatus, { AffiliateStatusKey } from "./affiliateStatus";
import {
  AffiliateType,
  affiliateTypeFromID,
  AffiliateTypeID,
  affiliateTypeToID,
} from "./affiliateType";
import AgencyConfiguration from "./agencyConfiguration";
import Agent from "./agent";
import AppRelease from "./appRelease";
import { AreaUnit, areaUnitFromID, AreaUnitID } from "./areaUnit";
import Category, {
  PopularCategoryIndex,
  PopularCategoryKey,
  popularCategoryKeyToIndex,
} from "./category";
import Client, { ClientAccessID, ClientReferralType, ClientStatusID, ClientTypeID } from "./client";
import ClientPhoneNumber from "./clientPhoneNumber";
import Contact from "./contact";
import Contract from "./contract";
import ContractListing from "./contractListing";
import ContractStatus from "./contractStatus";
import Country from "./country";
import Designation from "./designation";
import Lead, {
  LeadInquiryBudgetCurrency,
  LeadInquiryObjectType,
  LeadMarket,
  LeadSide,
  LeadState,
} from "./lead";
import {
  LeadClassification,
  leadClassificationFromID,
  LeadClassificationID,
  leadClassificationToID,
} from "./leadClassification";
import LeadLocation, { LeadLocationType } from "./leadLocation";
import LeadSource, { LeadSourceKey } from "./leadSource";
import LeadSourceMapping from "./leadSourceMapping";
import LeadStatus from "./leadStatus";
import Listing, {
  ListingAgainstType,
  ListingAllocationID,
  ListingCommissionType,
  ListingStatusID,
  ListingVisibilityID,
} from "./listing";
import ListingImage, { ListingImageCategory } from "./listingImage";
import ListingReservation from "./listingReservation";
import ListingStatus from "./listingStatus";
import Location, { LocationLevel, LocationLevelTitle, LocationPopularityIndex } from "./location";
import MediaBankImage from "./mediaBankImage";
import Platform from "./platform";
import PotentialAffiliate, {
  GovIDType,
  PotentialAffiliateContactData as PotentialAffiliateContactDataType,
  PotentialAffiliateDeviceData as PotentialAffiliateDeviceDataType,
  PotentialAffiliateStatus,
  PotentialAffiliateVerificationStatusID,
  PotentialAffiliateVerificationStatusKey,
} from "./potentialAffiliate";
import PotentialAffiliateVerificationStatus from "./potentialAffiliateVerificationStatus";
import Project, { ProjectAffiliateVisibilityID, ProjectStatus } from "./project";
import ProjectImage from "./projectImage";
import { Purpose, purposeFromID, PurposeID, purposeToID } from "./purpose";
import Task, { TaskAgainstType, TaskStageID, TaskStatus } from "./task";
import TaskType from "./taskType";
import User from "./user";

const sequelizeNamespace = cls.createNamespace("affiliates-api");
Sequelize.useCLS(sequelizeNamespace);

// @ts-ignore because typings don't include `operatorsAliases`
const sequelize = new Sequelize(config.get("db").url, {
  operatorsAliases: 0,
  logging: false,
  timezone: getDbTimezone(),
});

Address.setup(sequelize);
Affiliate.setup(sequelize);
AffiliateStatus.setup(sequelize);
AffiliateAssignee.setup(sequelize);
AffiliateContact.setup(sequelize);
AffiliateContactCell.setup(sequelize);
AgencyConfiguration.setup(sequelize);
AffiliateCommission.setup(sequelize);
AffiliateCommissionPropertyType.setup(sequelize);
Agent.setup(sequelize);
Category.setup(sequelize);
Contact.setup(sequelize);
Client.setup(sequelize);
ClientPhoneNumber.setup(sequelize);
Country.setup(sequelize);
Contract.setup(sequelize);
ContractListing.setup(sequelize);
ContractStatus.setup(sequelize);
Designation.setup(sequelize);
Lead.setup(sequelize);
LeadLocation.setup(sequelize);
LeadStatus.setup(sequelize);
LeadSource.setup(sequelize);
LeadSourceMapping.setup(sequelize);
Listing.setup(sequelize);
ListingStatus.setup(sequelize);
ListingImage.setup(sequelize);
MediaBankImage.setup(sequelize);
Location.setup(sequelize);
PotentialAffiliate.setup(sequelize);
PotentialAffiliateVerificationStatus.setup(sequelize);
Project.setup(sequelize);
ProjectImage.setup(sequelize);
Task.setup(sequelize);
TaskType.setup(sequelize);
User.setup(sequelize);
ListingReservation.setup(sequelize);
AppRelease.setup(sequelize);
Platform.setup(sequelize);

AffiliateCommission.belongsTo(AffiliateCommissionPropertyType, { foreignKey: "propertyTypeId" });

AffiliateStatus.hasMany(Affiliate, {
  sourceKey: "id",
  foreignKey: "affiliateStatus",
});
Address.hasMany(Project, {
  foreignKey: "addressID",
  sourceKey: "id",
});
Address.belongsTo(Location, { foreignKey: "cityLocationID" });

AffiliateAssignee.hasOne(User, { sourceKey: "userID", foreignKey: "id", constraints: false });
AffiliateAssignee.hasOne(Affiliate, {
  sourceKey: "affiliateID",
  foreignKey: "id",
  constraints: false,
});

Affiliate.hasOne(AffiliateContact, { sourceKey: "id", foreignKey: "affiliateID" });
Affiliate.hasMany(AffiliateAssignee, {
  sourceKey: "id",
  foreignKey: "affiliateID",
});
Affiliate.hasOne(AffiliateStatus, {
  sourceKey: "affiliateStatus",
  foreignKey: "id",
  constraints: false,
});

AffiliateContact.belongsTo(Affiliate, { foreignKey: "affiliateID" });

AffiliateContact.belongsTo(Contact, { foreignKey: "contactID", constraints: false });

Contact.hasOne(AffiliateContactCell, { foreignKey: "contactID" });

AffiliateAssignee.belongsTo(AffiliateContact, { foreignKey: "affiliateID" });

Affiliate.belongsTo(Client, { foreignKey: "referralID" });
Affiliate.belongsTo(Agent, { foreignKey: "agencyID" });

Client.hasOne(AffiliateContact, { sourceKey: "referralID", foreignKey: "affiliateID" });
AffiliateContact.belongsTo(Client, { foreignKey: "referralID" });

Client.hasMany(ClientPhoneNumber, { foreignKey: "clientID" });

Client.belongsTo(Location, {
  foreignKey: {
    name: "locationID",
    allowNull: true,
  },
});

Client.belongsTo(Country, {
  foreignKey: {
    name: "countryID",
    allowNull: true,
  },
});

Category.hasMany(Listing, {
  foreignKey: { name: "categoryID", allowNull: true },
  sourceKey: "id",
});

ContractListing.belongsTo(Contract, { foreignKey: "contractId" });

Contract.belongsTo(ContractStatus, { foreignKey: "status" });

PotentialAffiliate.hasOne(AffiliateContact, {
  sourceKey: "affiliateContactID",
  foreignKey: {
    name: "affiliateContactID",
    allowNull: true,
  },
  constraints: false,
});

PotentialAffiliate.belongsTo(PotentialAffiliateVerificationStatus, {
  foreignKey: "verificationStatusID",
});

Lead.belongsTo(Category, {
  foreignKey: {
    name: "inquiryCategoryID",
    allowNull: true,
  },
});

Lead.belongsTo(User, { foreignKey: "userID" });
Lead.belongsTo(LeadStatus, { foreignKey: "statusID" });
Lead.belongsTo(Client, { foreignKey: "clientID" });
Lead.belongsTo(Address, {
  foreignKey: {
    name: "addressID",
    allowNull: true,
  },
});

Lead.hasMany(LeadLocation, { foreignKey: "leadID" });

Lead.hasOne(Task, {
  foreignKey: "againstID",
  sourceKey: "id",
  as: "LastTaskPerformed",
});

LeadLocation.belongsTo(Location, { foreignKey: "locationID" });

Task.belongsTo(User, { foreignKey: "addedByUserID", as: "AddedByUser" });
Task.belongsTo(User, { foreignKey: "assignedToUserID", as: "AssignedToUser" });

Task.belongsTo(TaskType, { foreignKey: "typeID" });
TaskType.belongsTo(TaskType, { foreignKey: "parentID", as: "ParentTaskType" });

Project.hasMany(ProjectImage, { foreignKey: "projectID" });

Project.hasMany(Listing, {
  foreignKey: "againstID",
  sourceKey: "id",
  scope: { againstType: "project" },
});

Project.belongsTo(Address, { foreignKey: "addressID", targetKey: "id" });

Listing.belongsTo(Project, { foreignKey: "againstID", targetKey: "id" });

Listing.belongsTo(ListingStatus, {
  foreignKey: "statusID",
});

Listing.belongsTo(Category, { foreignKey: "categoryID" });

Listing.belongsTo(Address, {
  foreignKey: {
    name: "addressID",
    allowNull: true,
  },
});

Listing.hasMany(ContractListing, {
  foreignKey: "objectId",
  scope: { objectType: "listing" },
});

Location.belongsTo(Country, {
  foreignKey: {
    name: "countryID",
    allowNull: true,
  },
});

Location.belongsTo(Location, {
  foreignKey: {
    name: "level1LocationID",
    allowNull: true,
  },
  as: "Level1Location",
});

Listing.hasMany(ListingImage, { foreignKey: "listingID" });

ListingImage.belongsTo(MediaBankImage, {
  foreignKey: "mediaBankImageID",
});
ListingReservation.belongsTo(Affiliate, {
  foreignKey: "reservedBy",
  as: "ReservedByAffiliate",
});
ListingReservation.belongsTo(User, {
  foreignKey: "reservedBy",
  as: "ReservedBy",
});
ListingReservation.belongsTo(User, { foreignKey: "reservedFor", as: "ReservedFor" });
ListingReservation.belongsTo(User, { foreignKey: "updatedBy", as: "UpdatedBy" });
ListingReservation.belongsTo(Lead, { foreignKey: "leadID" });
ListingReservation.belongsTo(Listing, { foreignKey: "listingID" });

export type PotentialAffiliateContactData = PotentialAffiliateContactDataType;
export type PotentialAffiliateDeviceData = PotentialAffiliateDeviceDataType;

export {
  sequelize,
  sequelizeNamespace,
  Address,
  Affiliate,
  AffiliatePlatform,
  AffiliatePlatformID,
  affiliatePlatformToID,
  AffiliateStatus,
  AffiliateStatusKey,
  AffiliateAssignee,
  AffiliateAssigneeType,
  AffiliateCommission,
  AffiliateCommissionPropertyType,
  AffiliateContact,
  AffiliateContactCell,
  AgencyConfiguration,
  AffiliateType,
  AffiliateTypeID,
  affiliateTypeFromID,
  affiliateTypeToID,
  Agent,
  AreaUnit,
  AreaUnitID,
  AppRelease,
  areaUnitFromID,
  Category,
  Contact,
  Client,
  ClientStatusID,
  ClientTypeID,
  ClientAccessID,
  ClientReferralType,
  ClientPhoneNumber,
  Country,
  Contract,
  ContractListing,
  Designation,
  Location,
  Lead,
  LeadState,
  LeadSide,
  LeadMarket,
  LeadInquiryObjectType,
  LeadInquiryBudgetCurrency,
  LeadLocation,
  LeadSource,
  LeadSourceKey,
  LeadSourceMapping,
  LeadLocationType,
  LeadClassification,
  LeadClassificationID,
  leadClassificationFromID,
  leadClassificationToID,
  LeadStatus,
  Listing,
  ListingAgainstType,
  ListingStatusID,
  ListingVisibilityID,
  ListingAllocationID,
  ListingCommissionType,
  ListingImage,
  ListingImageCategory,
  MediaBankImage,
  PotentialAffiliate,
  PotentialAffiliateVerificationStatus,
  PotentialAffiliateStatus,
  PotentialAffiliateVerificationStatusID,
  PotentialAffiliateVerificationStatusKey,
  GovIDType,
  Project,
  ProjectStatus,
  ProjectAffiliateVisibilityID,
  ProjectImage,
  Purpose,
  PurposeID,
  purposeFromID,
  purposeToID,
  Task,
  TaskStatus,
  TaskStageID,
  TaskAgainstType,
  TaskType,
  User,
  ListingReservation,
  LocationPopularityIndex,
  LocationLevel,
  LocationLevelTitle,
  ListingStatus,
  PopularCategoryKey,
  PopularCategoryIndex,
  popularCategoryKeyToIndex,
  ContractStatus,
  Platform,
};
