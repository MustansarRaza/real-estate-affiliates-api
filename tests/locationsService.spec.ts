import { Country, Location } from "affiliates-api/models";
import { LocationsService } from "affiliates-api/services";
import { loadSequelizeFixtures } from "./support";
import { ModelFixtureNames } from "./support/data";

describe("LocationsService", () => {
  const service = new LocationsService();

  describe("list", () => {
    it("orders by level and returns type ID based on level", async () => {
      const country = await Location.create({
        name: "Netherlands",
        parentID: null,
        level: 1,
      });

      const region = await Location.create({
        name: "Gelderland",
        parentID: country.id,
        level: 2,
      });

      const city = await Location.create({
        name: "Ede",
        parentID: region.id,
        level: 3,
      });

      const locations = await service.list();
      expect(locations).toMatchObject([
        {
          id: country.id,
          name: country.name,
          parentID: null,
          typeID: 0,
        },
        {
          id: region.id,
          name: region.name,
          parentID: country.id,
          typeID: 1,
        },
        {
          id: city.id,
          name: city.name,
          parentID: region.id,
          typeID: 2,
        },
      ]);
    });

    it("returns undefined for typeID if not available", async () => {
      const loc = await Location.create({
        name: "Random",
        parentID: null,
        level: 4,
      });

      const locations = await service.list();
      expect(locations).toHaveLength(1);
      expect(locations[0].id).toBe(loc.id);
      expect(locations[0].typeID).toBe(undefined);
    });

    it("returns undefined for iso31662Code if not available", async () => {
      const loc = await Location.create({
        name: "Pakistan",
        parentID: null,
        level: 1,
      });

      const locations = await service.list();
      expect(locations).toHaveLength(1);
      expect(locations[0].id).toBe(loc.id);
      expect(locations[0].iso31662Code).toBe(undefined);
    });

    it("returns undefined for iso31662Code if not a country", async () => {
      const country = await Country.create({
        name: "Pakistan",
        iso31662Code: "PK",
      });

      // is in Pakistan, but is not a country node, should
      // not get a ISO31662 code
      const loc = await Location.create({
        name: "Random",
        parentID: null,
        level: 4,
        countryID: country.id,
      });

      const locations = await service.list();
      expect(locations).toHaveLength(1);
      expect(locations[0].id).toBe(loc.id);
      expect(locations[0].iso31662Code).toBe(undefined);
    });

    it("returns ISO31662 code for country level nodes", async () => {
      const country = await Country.create({
        name: "Pakistan",
        iso31662Code: "PK",
      });

      const loc = await Location.create({
        name: "Pakistan",
        parentID: null,
        countryID: country.id,
        level: 1,
      });

      const locations = await service.list();
      expect(locations).toHaveLength(1);
      expect(locations[0].id).toBe(loc.id);
      expect(locations[0].iso31662Code).toBe("PK");
    });
  });

  describe("getWithHierarchy", () => {
    it("gets a location with hierarchy data", async () => {
      const countryLocation = await Location.create({
        name: "Pakistan",
        parentID: null,
        level: 1,
      });

      const location = await Location.create({
        name: "Punjab",
        parentID: 1,
        level: 2,
        level1LocationID: countryLocation.id,
      });

      const dto = await service.getWithHierarchy({ locationID: location.id });
      expect(dto).toMatchSnapshot();
    });

    it("returns null if no such location exists", async () => {
      const dto = await service.getWithHierarchy({ locationID: 999 });
      expect(dto).toBe(null);
    });
  });

  describe("getlocationCityIDMapping", () => {
    it("gets a Map having cityId against the location", async () => {
      await loadSequelizeFixtures(ModelFixtureNames.COUNTRIES, ModelFixtureNames.LOCATIONS);

      const location = await Location.create({
        name: "Gajju Matta",
        parentID: 3,
        level: 4,
        level1LocationID: 1,
        level3LocationID: 3,
      });

      const locationCityIdMap = await service.getLocationToCityIDMapping([location.id]);
      expect(locationCityIdMap.size).toEqual(1);
      expect(locationCityIdMap.get(location.id)).toEqual(3);
    });

    it("returns empty mapping if there is no locationId provided", async () => {
      await loadSequelizeFixtures(ModelFixtureNames.COUNTRIES, ModelFixtureNames.LOCATIONS);

      const locationCityIdMap = await service.getLocationToCityIDMapping([]);
      expect(locationCityIdMap.size).toBe(0);
    });
  });

  describe("countryID", () => {
    it("returns the countryID for a locationID", async () => {
      const country = await Country.create({
        name: "Netherlands",
      });

      const location = await Location.create({
        name: "Ede",
        parentID: null,
        countryID: country.id,
        level: 1,
      });

      const countryID = await service.countryID({ locationID: location.id });
      expect(countryID).toBe(country.id);
    });

    it("returns null if the location has no countryID", async () => {
      const location = await Location.create({
        name: "Ede",
        parentID: null,
        countryID: null,
        level: 1,
      });

      const countryID = await service.countryID({ locationID: location.id });
      expect(countryID).toBe(null);
    });

    it("throws an error if the location could not be found", async () => {
      const countryID = await service.countryID({ locationID: 1337 });
      expect(countryID).toBe(null);
    });
  });
});
