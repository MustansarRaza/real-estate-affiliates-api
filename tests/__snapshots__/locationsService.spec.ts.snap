// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationsService getWithHierarchy gets a location with hierarchy data 1`] = `
Object {
  "id": 2,
  "iso31662Code": undefined,
  "level1Location": Object {
    "id": 1,
    "iso31662Code": undefined,
    "name": "Pakistan",
    "parentID": null,
    "popularityIndex": undefined,
    "typeID": 0,
  },
  "name": "Punjab",
  "parentID": 1,
  "popularityIndex": undefined,
  "typeID": 1,
}
`;
